File: 2164-two-best-non-overlapping-events/2164-two-best-non-overlapping-events.java
import java.util.*;

class Solution {
    public int maxTwoEvents(int[][] events) {
        List<int[]> time = new ArrayList<>();
        
        for (int[] event : events) {
            int st = event[0];
            int et = event[1];
            int val = event[2];
            
            time.add(new int[] {st, 0, val});
            time.add(new int[] {et, 1, val});
        }
        
        // Sort the timeline: by time, and for same time, prioritize end times (1) over start times (0)
        Collections.sort(time, (a, b) -> a[0] != b[0] ? Integer.compare(a[0], b[0]) : Integer.compare(a[1], b[1]));
        
        int prevMax = 0, ans = 0;
        
        for (int[] event : time) {
            int t = event[0];
            int end = event[1];
            int val = event[2];
            
            if (end == 1) {
                prevMax = Math.max(prevMax, val);
            } else {
                ans = Math.max(ans, prevMax + val);
            }
        }
        
        return ans;
    }
}

File: 2164-two-best-non-overlapping-events/2164-two-best-non-overlapping-events.py
class Solution:
    def maxTwoEvents(self, events: List[List[int]]) -> int:
        time = []
        
        for event in events:
            st, et, val = event
            time.append((st, 0, val))  # 0 indicates start
            time.append((et, 1, val))  # 1 indicates end
        
        # Sort the timeline: by time, and for the same time, prioritize end times (1) over start times (0)
        time.sort(key=lambda x: (x[0], x[1]))
        
        prev_max = 0
        ans = 0
        
        for t, is_end, val in time:
            if is_end:  # If it's an end time
                prev_max = max(prev_max, val)
            else:  # If it's a start time
                ans = max(ans, prev_max + val)
        
        return ans
